evalml/pipelines/component_graph.py

```diff
@@ -414,7 +420,10 @@ def _transform_features(
             self.input_feature_names.update({component_name: list(x_inputs.columns)})
             if isinstance(component_instance, Transformer):
                 if fit:
-                    output = component_instance.fit_transform(x_inputs, y_input)
+                    if not component_instance._is_fitted or self.cached_data is None:
```

Comment:
@chukarsten before introducing a cache, if we call `fit`, we would fit/refit all components in the component graph. There were no scenarios here where calling fit could allow us to skip fitting certain components. However, now, there are things we want to keep in mind:
1) If we call fit and the `component_instance` in question is already fitted but there is no cached data, then we want to refit. This can happen in cases where we pull an already-fitted pipeline and call fit on some `X, y` data. Without knowing whether or not we've fitted originally on this `X,  y` data, we should opt to re-fit this, even if the component instance was already fitted.
2) If the `component_instance` is already fitted and we have cached data, we can instead just rely on the cached data, as long as the hashes match. We only pass in cached data during creation of the ensembler, so this should only trigger when we are using ensembling inside of `AutoMLSearch`. 

The logic above handles these scenarios to ensure that only when we have cached data and have an already-fitted component, can we skip the `fit` call. Otherwise, it's best to call `fit` again to ensure the pipelines can operate as expected.

Reply:
So basically we only choose to fit skipping if the instance is fitted AND we have cached data right? Maybe it could be helpful to swap the if/else here to make it more clear what case we're not fitting?

Reply:
@bchen1116 

I have a question about 

> This can happen in cases where we pull an already-fitted pipeline and call fit on some X, y data. Without knowing whether or not we've fitted originally on this X, y data

If `get_relevant_component` gets the components from the cache corresponding to the hash of the data, and the `X`, and `y` in question are not in the cache, how will the component instance be fitted?

Reply:
Good point! A smooth-brained moment. Had made some code changes that makes this double-check irrelevant